File: src/UnwindMC/Analysis/ImportResolver.cs
using System;
using System.Collections.Generic;
using UnwindMC.Util;

namespace UnwindMC.Analysis
{
    public class ImportResolver
    {
        private readonly ulong _imageBase;
        private readonly ArraySegment<byte> _importAddressTableBytes;
        private readonly ArraySegment<byte> _importBytes;
        private readonly Dictionary<int, string> _imports = new Dictionary<int, string>();

        public ImportResolver(ulong imageBase, ArraySegment<byte> importAddressTableBytes, ArraySegment<byte> importBytes)
        {
            _imageBase = imageBase;
            _importAddressTableBytes = importAddressTableBytes;
            _importBytes = importBytes;
        }

        public bool IsImportAddress(ulong address)
        {
            var virtualAddress = address - _imageBase;
            return virtualAddress >= (uint)_importAddressTableBytes.Offset &&
                virtualAddress <= (uint)(_importAddressTableBytes.Offset + _importAddressTableBytes.Count - 4);
        }

        public string GetImportName(ulong address)
        {
            uint entryAddress = _importAddressTableBytes.Array.ReadUInt32((int)(address - _imageBase));
            ushort hint = _importBytes.Array.ReadUInt16((int)entryAddress);
            string name;
            if (!_imports.TryGetValue(hint, out name))
            {
                name = _importBytes.Array.ReadZString((int)entryAddress + 2);
                _imports[hint] = name;
            }
            return name;
        }
    }
}

File: src/UnwindMC/Analysis/Imports/ImportResolver.fs
module ImportResolver

open System
open System.Collections.Generic
open IImportResolver

type ImportResolver(imageBase: uint64, importAddressTableBytes: ArraySegment<byte>, importBytes: ArraySegment<byte>) =
    let _imageBase: uint64 = imageBase
    let _importAddressTableBytes: ArraySegment<byte> = importAddressTableBytes
    let _importBytes: ArraySegment<byte> = importBytes
    let _imports: Dictionary<int, string> = new Dictionary<int, string>()

    interface IImportResolver with
        member self.IsImportAddress(address: uint64): bool =
            let virtualAddress = address - _imageBase
            virtualAddress >= (uint64)_importAddressTableBytes.Offset &&
                virtualAddress <= (uint64)(_importAddressTableBytes.Offset + _importAddressTableBytes.Count - 4)

        member self.GetImportName(address: uint64): string =
            let entryAddress = ArrayReader.readUInt32 _importAddressTableBytes.Array ((int)(address - _imageBase))
            let hint = ArrayReader.readUInt16 _importBytes.Array ((int)entryAddress)
            let hasValue, name = _imports.TryGetValue((int)hint)
            if hasValue then
                let name = ArrayReader.readZString _importBytes.Array ((int)entryAddress + 2)
                _imports.[(int)hint] <- name;
                name
            else
                name
